File: game/bin/tools/process_defs/resources/scripts/GenerateCSharp/templates/EntityFactory.cs
using System;
using System.Collections.Generic;
using System.Text;

{% if generatedNamespace != "BW" -%}
using BW;{%- endif %}

namespace {{ generatedNamespace }}
{

/// <summary>
/// This class is responisible for entity creation off
/// entity type id
/// </summary>
class EntityFactory : IEntityFactory
{
	#region Constructors
	public EntityFactory()
	{
{%- for entity in entityDescriptions %}		
		creators_[ {{ entity.clientIndex}} ] =
			( ConnectionWithModel conn ) => 
			{ return new {{ entity.name }}{% if shouldGenerateView %}View{% endif %}( conn ); };
{% endfor %}
	}
	#endregion Constructors

	/// <summary>
	/// This method creates an entity
	/// </summary>
	/// <param name="entityTypeID">Entity Type ID</param>
	/// <param name="connection">ConnectionWithModel instance for the new entity</param>
	/// <returns></returns>
	public Entity create( Int16 entityTypeID, ConnectionWithModel connection )
	{
		DCreator creator;
		if (!creators_.TryGetValue( entityTypeID, out creator ))
		{
			return null;
		}
		return creator( connection );
	}

	#region Private helpers
	delegate Entity DCreator( ConnectionWithModel connection );
	Dictionary<Int16, DCreator> creators_ = new Dictionary<Int16, DCreator>();
	#endregion Private helpers
}


} // namespace {{ generatedNamespace }}

File: game/bin/tools/process_defs/resources/scripts/GenerateJavaScript/templates/EntityFactory.cpp
#include "EntityFactory.hpp"


{% for entity in entityDescriptions %}
#include "{{entityPath}}/{{ entity.name }}.hpp"{% endfor %}


/**
 *	This method creates the appropriate sub-class instance of BWEntity based on
 *	the given entity type ID.
 *
 *	@param typeID	The ID of the entity type to create.
 *	@param pBWConnection A pointer to the connection to the BigWorld server.
 *
 *	@return 	An instance of the appropriate sub-class of BWEntity, or NULL
 *				if the entity type ID is invalid.
 */
BW::BWEntity * EntityFactory::doCreate( BW::EntityTypeID entityTypeID,
		BW::BWConnection * pBWConnection )
{
	BW::BWEntity * pEntity = NULL;

	switch (entityTypeID)
	{
{% for entity in entityDescriptions %}		case {{ entity.clientIndex|rjust( "3" ) }}: pEntity = new {{ entity.name }}( pBWConnection ); break;
{% endfor %}	}

	return pEntity;
}


// EntityFactory.cpp

